client/neo4j.go

```diff
@@ -0,0 +1,151 @@
+package client
+
+import (
+	logKeys "github.com/ONSdigital/dp-dimension-importer/common"
+	"github.com/ONSdigital/go-ns/log"
+	bolt "github.com/johnnadratowski/golang-neo4j-bolt-driver"
+	"os"
+	"sync"
+	"errors"
+)
+
+//go:generate moq -out generated_bolt_mocks.go . NeoConn NeoDriverPool NeoStmt NeoQueryRows NeoResult
+
+// NeoConn type to easliy allow MOQ to generate a mock from the bolt.Conn interface
+type NeoConn bolt.Conn
+
+// NeoStmt type to easliy allow MOQ to generate a mock from the bolt.Stmt interface
+type NeoStmt bolt.Stmt
+
+type NeoQueryRows bolt.Rows
+
+type NeoResult bolt.Result
+
+// NeoDriverPool defines interface for bolt driver pool
+type NeoDriverPool interface {
+	OpenPool() (bolt.Conn, error)
+}
+
+type NewDriverPool func(connStr string, max int) (bolt.DriverPool, error)
+
+// Neo4j client provides functions for inserting dimension, instances and relationships into a Neo4j database.
+type Neo4j struct{}
+
+const (
+	stmtKey       = "statment"
+	stmtParamsKey = "params"
+
+	openConnErr          = "Error while attempting to open connection"
+	errExecutingStatment = "Error while attempting to execute statement"
+	errDrivePoolInit     = "Error while attempting to initialize neo4j driver pool"
+	errCreatingStatment  = "Error while attempting to create neo4j statement"
+	errQueryWasEmpty     = "Query is required but was nil or empty"
+	errStmtWasEmpty      = "Statement is required but was nil or empty"
+	errRetrievingRows    = "Error while attempting to retrieve row data"
+)
+
+var (
+	once          sync.Once
+	newDriverPool NewDriverPool = bolt.NewDriverPool
+	neoDriverPool NeoDriverPool
+)
+
+type NeoRows struct {
+	Data [][]interface{}
+}
+
+// NewDatabase creates a new instance of the Neo4j client and does the initial setup required for the
+// client to connect to the database instance.
+func NewDatabase(url string, poolSize int) Neo4j {
+	once.Do(func() {
+		pool, err := bolt.NewDriverPool(url, poolSize)
+		if err != nil {
+			log.ErrorC(errDrivePoolInit, err, log.Data{
+				logKeys.URL:      url,
+				logKeys.PoolSize: poolSize,
+			})
+			os.Exit(1)
+		}
+		neoDriverPool = pool
+	})
+	return Neo4j{}
+}
+
+func (neo Neo4j) Query(query string, params map[string]interface{}) (*NeoRows, error) {
+	if len(query) == 0 {
+		return nil, errors.New(errQueryWasEmpty)
+	}
+	logData := map[string]interface{}{
+		stmtKey:       query,
+		stmtParamsKey: params,
+	}
+	var conn bolt.Conn
+	var err error
+
+	if conn, err = neoDriverPool.OpenPool(); err != nil {
+		log.ErrorC(openConnErr, err, nil)
+		return nil, err
+	}
+	defer conn.Close()
+
+	var neoStmt bolt.Stmt
+	if neoStmt, err = createStmt(conn, query); err != nil {
```

Comment:
Iv tried to separate out all dimension / instance specific logic from the Neo4j struct, its only concern is executing statements & queries. I have the dimension & instance repositories that focus on the specifics of dimensions & instances.